cmd/tpm-manager/tpm-manager.go

```
@@ -0,0 +1,544 @@
+	}
+
+	// Don't handle this error immediately - we want to update the annotations even if validation failed
+	logstate, err := tpm.ValidatePCRs(log, quote, manager.pcrConfigs)
+	jsonlog, jsonerr := json.Marshal(logstate)
+
+	if jsonerr == nil {
+		updateAnnotations(node, string(jsonlog))
+	} else {
+		glog.Errorf("Failed to serialise new log state: %v", jsonerr)
+	}
+
+	if err != nil {
+		err = invalidateNode(node)
+		if err != nil {
+			return fmt.Errorf("Failed to invalidate node: %v", err)
+		}
+		return fmt.Errorf("Invalidating Node: Unable to validate quote for node %s", address.String())
+	}
+
+	// If we've got this far then the node is trustworthy
+	err = trustNode(node)
+	if err != nil {
+		return fmt.Errorf("Failed to mark node as trustworthy")
+	}
+	return nil
+}
+
+func verifyAndUpdate(node *api.Node) {
+	err := verifyNode(node)
+	if err != nil {
+		glog.Errorf("Failed to verify node %s: %v", node.Name, err)
+	}
+	// The state that the node will be updated to
+	newstate := isTrusted(node)
+	newnode, err := manager.client.Nodes().Get(node.Name)
+	if err != nil {
+		glog.Errorf("Unable to obtain node state for %s: %v", node.Name, err)
+		return
+	}
+	if newnode.ObjectMeta.Annotations == nil {
+		newnode.ObjectMeta.Annotations = make(map[string]string)
+	}
+	currenttime := time.Now().Unix()
+
+	// If we're transitioning state, update the metadata
+	if isTrusted(newnode) != newstate {
+		if newstate == false {
+			newnode.ObjectMeta.Annotations[UntrustedSince] = strconv.FormatInt(currenttime, 10)
+			newnode.ObjectMeta.Annotations[TrustedSince] = ""
+		} else {
+			newnode.ObjectMeta.Annotations[UntrustedSince] = ""
+			newnode.ObjectMeta.Annotations[TrustedSince] = strconv.FormatInt(currenttime, 10)
+		}
+	}
+
+	newnode.ObjectMeta.Annotations[ValidationTime] = strconv.FormatInt(currenttime, 10)
+
+	// Ensure that the new node is tainted appropriately
+	if newstate == true {
+		trustNode(newnode)
```

ignoring error return